app/src/test/java/de/westnordost/streetcomplete/quests/opening_hours/OpeningHoursTagParserTest.kt

```diff
@@ -0,0 +1,397 @@
+        Assert.assertEquals(returned[0].weekdaysList.size, 2)
+        Assert.assertEquals(returned[0].weekdaysList[0].timeRange.start, 9 * 60)
+        Assert.assertEquals(returned[0].weekdaysList[0].timeRange.end, 12 * 60)
+        Assert.assertEquals(returned[0].weekdaysList[0].timeRange.isOpenEnded, false)
+        Assert.assertEquals(returned[0].weekdaysList[1].timeRange.start, 13 * 60)
+        Assert.assertEquals(returned[0].weekdaysList[1].timeRange.end, 14 * 60)
+        Assert.assertEquals(returned[0].weekdaysList[1].timeRange.isOpenEnded, false)
+        val weekData = BooleanArray(7) { true }
+        Assert.assertEquals(returned[0].weekdaysList[0].weekdays, Weekdays(weekData))
+    }
+
+    @Test
+    fun `saving data to a tag without data loss`() {
+        Assert.assertEquals(OpeningHoursTagParser.internalIntoTag((OpeningHoursTagParser.parse("Mo 09:00-20:00")!!.toOpeningMonthsList())), "Mo 09:00-20:00")
+    }
+
+    @Test
+    fun `saving data to a tag without data loss also on multiple rules`() {
+        Assert.assertEquals(OpeningHoursTagParser.internalIntoTag((OpeningHoursTagParser.parse("Mo 09:00-20:00; Tu 00:00-24:00")!!.toOpeningMonthsList())), "Mo 09:00-20:00; Tu 00:00-24:00")
+    }
+
+    @Test
+    fun `accept additional whitespace and allow its removal`() {
+        Assert.assertEquals(OpeningHoursTagParser.internalIntoTag((OpeningHoursTagParser.parse("Tu-Su   10:30    -   18:00")!!.toOpeningMonthsList())), "Tu-Su 10:30-18:00")
+    }
+
+    @Test
+    fun `do not merge separate rules with the same hours`() {
+        Assert.assertEquals(OpeningHoursTagParser.internalIntoTag((OpeningHoursTagParser.parse("Mo 09:00-20:00; Tu 09:00-20:00")!!.toOpeningMonthsList())), "Mo 09:00-20:00; Tu 09:00-20:00")
+    }
+
+    @Test
+    fun `do not reorder rules`() {
+        Assert.assertEquals(OpeningHoursTagParser.internalIntoTag((OpeningHoursTagParser.parse("Mo 09:00-20:00; Su 10:00-11:00; Tu 09:00-20:00")!!.toOpeningMonthsList())), "Mo 09:00-20:00; Su 10:00-11:00; Tu 09:00-20:00")
+    }
+
+    @Test
+    fun `reject unimplemented for now 'till last client' variant`() {
+        Assert.assertEquals(OpeningHoursTagParser.parse("Mo 09:00-20:00+"), null)
+    }
+
+    @Test
+    fun `reject unimplemented for now pairs of days`() {
+        // note cases like "Mo,Tu,We 9:00-10:00" that will be either reject or transformed into "Mo-We 9:00-10:00"
+        Assert.assertEquals(OpeningHoursTagParser.parse("Mo,Tu 09:00-20:00"), null)
+    }
+
+    @Test
+    fun `reject unimplemented for now groups of days`() {
+        Assert.assertEquals(OpeningHoursTagParser.parse("Mo,Tu,Sa 09:00-20:00"), null)
+    }
+
+    @Test
+    fun `reject multiple lists of days in a single rule, as it would change form on processing, possibly with unexpected and unwanted results`() {
+        Assert.assertEquals(OpeningHoursTagParser.parse("Mo-Fr 7:30-18:00, Sa-Su 9:00-18:00"), null)
+    }
```

Comment:
As two separate rules. `Mo-Fr 7:30-18:00; Sa-Su 9:00-18:00`

In theory it would be possible to convert `Mo-Fr 7:30-18:00, Sa-Su 9:00-18:00` into `Mo-Fr 7:30-18:00; Sa-Su 9:00-18:00` but I went with rejecting this instead (at least for now) as it

- would get quite complex
- this syntax is fairly rare
- would be a structural, not a cosmetic change and I decided to go with "cosmetic changes only"

Reply:
> As two separate rules. Mo-Fr 7:30-18:00; Sa-Su 9:00-18:00

Ah, that's what you mean. Well, are you actually sure about this? I think the current state might be that StreetComplete uses the `;` unless the weekdays overlap.

Reply:
`Mo,Tu,Sa 09:00-20:00`, `Mo-Th,Sa 09:00-20:00`, `Mo-Th,PH 09:00-20:00` - nonintersecting ranges, shared hours are supported by SC and could be supported also by this quest.

`Mo-Fr 7:30-18:00, Sa-Su 9:00-18:00` makes sense but is not supported by SC, due to differing hours it is represented as two separate rules. Splitting such rules into multiple ones would work (as long as ranges are not intersecting etc) would work but for me it is changing structure what I would prefer to avoid, at least in the initial PR.

Reply:
I updated tests to describe this two situations separately - see https://github.com/westnordost/StreetComplete/pull/1676/commits/8096d23452d887da1ae3357ace53ba3842862639